lgc/patch/NggPrimShader.cpp

```diff
@@ -1372,7 +1428,53 @@ void NggPrimShader::constructPrimShaderWithoutGs(Module *module) {
     m_builder->SetInsertPoint(endExpPrimBlock);
 
     auto vertValid = m_builder->CreateICmpULT(m_nggFactor.threadIdInSubgroup, m_nggFactor.vertCountInSubgroup);
-    m_builder->CreateCondBr(vertValid, expVertBlock, endExpVertBlock);
+    if (disableCompact)
+      m_builder->CreateCondBr(vertValid, checkEmptyWaveBlock, endExpVertBlock);
+    else
+      m_builder->CreateCondBr(vertValid, expVertBlock, endExpVertBlock);
+  }
+
+  if (disableCompact) {
+    // Construct ".checkEmptyWave" block
+    {
+      m_builder->SetInsertPoint(checkEmptyWaveBlock);
+
+      auto emptyWave = m_builder->CreateICmpEQ(vertCountInWave, m_builder->getInt32(0));
+      m_builder->CreateCondBr(emptyWave, emptyWaveExpBlock, noEmptyWaveExpBlock);
+    }
+
+    // Construct ".emptyWaveExp" block
+    {
+      m_builder->SetInsertPoint(emptyWaveExpBlock);
+
+      auto firstThreadInWave = m_builder->CreateICmpEQ(m_nggFactor.threadIdInWave, m_builder->getInt32(0));
+      m_builder->CreateCondBr(firstThreadInWave, dummyExpVertBlock, endExpVertBlock);
+    }
```

Comment:
What's the motivation for this branch?

The one argument in favor that I could think of is that exporting data (moving from the WGP to the SX export block) costs energy, and moving only a single lane's worth of data costs less power than moving 32 lanes. But then, we create an export instruction with en==0, so none of the VGPRs should be exported at all. Which would make the branch unnecessary.

Is there another reason for doing the branch that I'm missing?